// ui5-testcafe-selector-utils/dist/ui5Builder.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ui5 = exports.UI5ObjectAttributeSelection = exports.UI5TableRowChainSelection = exports.UI5ListChainSelection = exports.UI5ComboBoxChainSelection = exports.UI5CoreItemSelection = exports.UI5MessageToast = exports.UI5ChainSelection = exports.UI5BaseBuilder = exports.UI5AnyValueBuilder = exports.UI5BaseBuilderIntf = void 0;
var testcafe_1 = require("testcafe");
var ui5Selector_1 = require("./ui5Selector");
var ui5Asserts_1 = require("./ui5Asserts");
var UI5BaseBuilderIntf = /** @class */ (function () {
    function UI5BaseBuilderIntf() {
        this._domQuery = "";
        this._name = "";
        this._hasTimeout = false;
        this._trace = false;
    }
    return UI5BaseBuilderIntf;
}());
exports.UI5BaseBuilderIntf = UI5BaseBuilderIntf;
;
var UI5AnyValueBuilder = /** @class */ (function (_super) {
    __extends(UI5AnyValueBuilder, _super);
    function UI5AnyValueBuilder(val) {
        var _this = _super.call(this) || this;
        _this._value = val;
        return _this;
    }
    UI5AnyValueBuilder.prototype.hasOwnTimeout = function () { return false; };
    UI5AnyValueBuilder.prototype.format = function () { return ""; };
    UI5AnyValueBuilder.prototype.selectAll = function () { return ""; };
    UI5AnyValueBuilder.prototype.trace = function () { return ""; };
    UI5AnyValueBuilder.prototype.isTraced = function () { return false; };
    UI5AnyValueBuilder.prototype.build = function (bInteractRequired) { return testcafe_1.Selector(""); };
    UI5AnyValueBuilder.prototype.property = function (propertyName, propertyValue) { return this; };
    UI5AnyValueBuilder.prototype.interactable = function () { return this; };
    // @ts-ignore
    UI5AnyValueBuilder.prototype.data = function (f) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this._value];
            });
        });
    };
    UI5AnyValueBuilder.prototype.dataSync = function (f) {
        return this._value;
    };
    UI5AnyValueBuilder.prototype.comboBox = function () { return new UI5ComboBoxChainSelection(this); };
    UI5AnyValueBuilder.prototype.parent = function (parent) { return this; };
    ;
    return UI5AnyValueBuilder;
}(UI5BaseBuilderIntf));
exports.UI5AnyValueBuilder = UI5AnyValueBuilder;
var UI5BaseBuilder = /** @class */ (function (_super) {
    __extends(UI5BaseBuilder, _super);
    function UI5BaseBuilder(chain) {
        var _this = _super.call(this) || this;
        _this.thisPointer = _this.getThisPointer();
        _this._id = {};
        if (chain && chain instanceof UI5BaseBuilderIntf) {
            _this._id = chain._id;
            _this._name = chain._name;
            _this._domQuery = chain._domQuery;
            _this._hasTimeout = chain._hasTimeout;
            _this._timeout = chain._timeout;
            _this._trace = chain._trace;
        }
        else if (chain && typeof chain == "string") {
            _this._name = chain;
        }
        _this._domQuery = chain ? chain._domQuery : null;
        return _this;
    }
    /** elements */
    UI5BaseBuilder.prototype.button = function () {
        return this.element("sap.m.Button");
    };
    UI5BaseBuilder.prototype.link = function () {
        return this.element("sap.m.Link");
    };
    UI5BaseBuilder.prototype.text = function () {
        return this.element("sap.m.Text");
    };
    UI5BaseBuilder.prototype.multiInput = function () {
        return this.element("sap.m.MultiInput");
    };
    UI5BaseBuilder.prototype.listItem = function () {
        return this.element("sap.m.ListItemBase");
    };
    UI5BaseBuilder.prototype.genericTile = function () {
        return this.element("sap.m.GenericTile");
    };
    UI5BaseBuilder.prototype.coreItem = function () {
        return new UI5CoreItemSelection(this.element("sap.ui.core.Item"));
    };
    UI5BaseBuilder.prototype.objectAttribute = function () {
        return new UI5ObjectAttributeSelection(this.element("sap.m.ObjectAttribute"));
    };
    UI5BaseBuilder.prototype.comboBox = function () {
        return new UI5ComboBoxChainSelection(this.element(["sap.m.MultiComboBox", "sap.m.ComboBox", "sap.m.Select"]));
    };
    UI5BaseBuilder.prototype.list = function () {
        return new UI5ListChainSelection(this.element("sap.m.List"));
    };
    UI5BaseBuilder.prototype.tableRow = function () {
        return new UI5TableRowChainSelection(this.element("sap.ui.table.Row").insideATable());
    };
    UI5BaseBuilder.prototype.messageToast = function () {
        return new UI5MessageToast(this);
    };
    UI5BaseBuilder.prototype.element = function (sElement) {
        this._id = this._enhanceWith(this._id, {
            metadata: {
                elementName: sElement
            }
        });
        return this.thisPointer;
    };
    UI5BaseBuilder.prototype.interactable = function () {
        this._id = this._enhanceWith(this._id, {
            metadata: {
                interactable: {
                    interactable: true
                }
            }
        });
        return this.thisPointer;
    };
    /** attributes */
    UI5BaseBuilder.prototype.bindingContextPath = function (modelName, path) {
        var oPath = {};
        oPath[modelName] = path;
        this._id = this._enhanceWith(this._id, {
            bindingContext: oPath
        });
        return this.thisPointer;
    };
    UI5BaseBuilder.prototype.bindingPath = function (attribute, path) {
        var oPath = {};
        oPath[attribute] = path;
        this._id = this._enhanceWith(this._id, {
            binding: oPath
        });
        return this.thisPointer;
    };
    UI5BaseBuilder.prototype.itemdata = function (itemPropertyName, itemPropertyValue) {
        var oPath = {};
        oPath[itemPropertyName] = itemPropertyValue;
        this._id = this._enhanceWith(this._id, {
            itemdata: {
                property: oPath
            }
        });
        return this.thisPointer;
    };
    UI5BaseBuilder.prototype.property = function (propertyName, propertyValue) {
        var oPath = {};
        oPath[propertyName] = propertyValue;
        this._id = this._enhanceWith(this._id, {
            property: oPath
        });
        return this.thisPointer;
    };
    UI5BaseBuilder.prototype.parentId = function (id) {
        this._id = this._enhanceWith(this._id, {
            parentAnyLevel: {
                identifier: {
                    id: id
                }
            }
        });
        return this.thisPointer;
    };
    UI5BaseBuilder.prototype.context = function (path, value) {
        var oProp = {};
        oProp[path] = value;
        this._id = this._enhanceWith(this._id, {
            smartContext: oProp
        });
        return this.thisPointer;
    };
    UI5BaseBuilder.prototype.parentProperty = function (name, value) {
        var oProp = {};
        oProp[name] = value;
        this._id = this._enhanceWith(this._id, {
            parentAnyLevel: {
                property: oProp
            }
        });
        return this.thisPointer;
    };
    UI5BaseBuilder.prototype.parentElementName = function (elem) {
        this._id = this._enhanceWith(this._id, {
            parentAnyLevel: {
                identifier: {
                    elementName: elem
                }
            }
        });
        return this.thisPointer;
    };
    UI5BaseBuilder.prototype.domChildWith = function (sDomWith) {
        this._id = this._enhanceWith(this._id, {
            domChildWith: sDomWith
        });
        return this.thisPointer;
    };
    UI5BaseBuilder.prototype.localViewName = function (viewName) {
        this._id = this._enhanceWith(this._id, {
            viewProperty: { localViewName: viewName }
        });
        return this.thisPointer;
    };
    UI5BaseBuilder.prototype.row = function (tableRow) {
        this._id = this._enhanceWith(this._id, {
            tableSettings: {
                insideATable: true,
                tableRow: tableRow
            }
        });
        return this.thisPointer;
    };
    UI5BaseBuilder.prototype.column = function (tableCol) {
        this._id = this._enhanceWith(this._id, {
            tableSettings: {
                insideATable: true,
                tableCol: tableCol
            }
        });
        return this.thisPointer;
    };
    UI5BaseBuilder.prototype.columnDescr = function (tableCol) {
        this._id = this._enhanceWith(this._id, {
            tableSettings: {
                insideATable: true,
                tableColDescr: tableCol
            }
        });
        return this.thisPointer;
    };
    UI5BaseBuilder.prototype.columnId = function (tableCol) {
        this._id = this._enhanceWith(this._id, {
            tableSettings: {
                insideATable: true,
                tableColId: tableCol
            }
        });
        return this.thisPointer;
    };
    UI5BaseBuilder.prototype.insideATable = function (tableRow, tableCol) {
        this._id = this._enhanceWith(this._id, {
            tableSettings: {
                insideATable: true,
                tableRow: tableRow,
                tableCol: tableCol
            }
        });
        return this.thisPointer;
    };
    UI5BaseBuilder.prototype.selectAll = function () {
        this._id = this._enhanceWith(this._id, {
            selectAll: true
        });
        return this.thisPointer;
    };
    UI5BaseBuilder.prototype.parent = function (parent) {
        if (parent._id && parent._id.identifier && parent._id.identifier.id) {
            this.parentId(parent._id.identifier.id);
        }
        if (parent._id && parent._id.metadata && parent._id.metadata.elementName) {
            this.parentElementName(parent._id.metadata.elementName);
        }
        return this.thisPointer;
    };
    UI5BaseBuilder.prototype.id = function (id) {
        this._id = this._enhanceWith(this._id, {
            identifier: {
                id: id
            }
        });
        return this.thisPointer;
    };
    UI5BaseBuilder.prototype.domQuery = function (id) {
        this._domQuery = id;
        return this.thisPointer;
    };
    UI5BaseBuilder.prototype.timeout = function (iTimeout) {
        this._timeout = iTimeout;
        this._hasTimeout = true;
        return this.thisPointer;
    };
    UI5BaseBuilder.prototype.hasOwnTimeout = function () {
        return this._hasTimeout;
    };
    UI5BaseBuilder.prototype.name = function (id) {
        this._name = id;
        return this.thisPointer;
    };
    UI5BaseBuilder.prototype.format = function () {
        if (this._domQuery) {
            return this._domQuery;
        }
        var sName = this._name + " ( " + JSON.stringify(this._id) + " )";
        return sName;
    };
    /** actions */
    UI5BaseBuilder.prototype.data = function (f) {
        return __awaiter(this, void 0, void 0, function () {
            var oSelector;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.build()];
                    case 1:
                        _a.sent(); //first wait until we generally see the element..
                        oSelector = this.build();
                        if (typeof oSelector.getUI5 !== "undefined") {
                            if (f) {
                                return [2 /*return*/, oSelector.getUI5(f)];
                            }
                            return [2 /*return*/, oSelector.getUI5()];
                        }
                        return [2 /*return*/, {}];
                }
            });
        });
    };
    UI5BaseBuilder.prototype.dataSync = function (f) {
        var oSelector = this.build();
        if (typeof oSelector.getUI5 !== "undefined") {
            return oSelector.getUI5(f);
        }
        return new Promise(function (resolve) {
            resolve({});
        });
    };
    /** expects */
    UI5BaseBuilder.prototype.expectProperty = function (property) {
        return this.expect().property(property);
    };
    UI5BaseBuilder.prototype.expectVisible = function (bExpectInteractable) {
        if (bExpectInteractable === void 0) { bExpectInteractable = true; }
        return this.expect().visible(bExpectInteractable);
    };
    UI5BaseBuilder.prototype.expectExists = function () {
        return this.expect().exists();
    };
    UI5BaseBuilder.prototype.expectDynamic = function (prop) {
        return this.expect().dynamic(prop);
    };
    UI5BaseBuilder.prototype.expect = function () {
        return ui5Asserts_1.ui5Assert(this);
    };
    /** helpers.. */
    UI5BaseBuilder.prototype.build = function (bInteractRequired, timeout) {
        if (this._domQuery) {
            return testcafe_1.Selector(this._domQuery);
        }
        if (bInteractRequired === true) {
            //i want to interact with this page element --> add attributes visible && enabled..
            this.interactable();
        }
        //hmmm.. probably not a good idea ?
        return ui5Selector_1.UI5Selector(timeout)(this._id);
    };
    UI5BaseBuilder.prototype.trace = function () {
        this._trace = true;
        return this.thisPointer;
    };
    UI5BaseBuilder.prototype.isTraced = function () {
        return this._trace;
    };
    UI5BaseBuilder.prototype._isObject = function (item) {
        return (item && typeof item === 'object' && !Array.isArray(item));
    };
    UI5BaseBuilder.prototype._mergeDeep = function (target) {
        var _a, _b;
        var sources = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            sources[_i - 1] = arguments[_i];
        }
        if (!sources.length)
            return target;
        var source = sources.shift();
        if (this._isObject(target) && this._isObject(source)) {
            for (var key in source) {
                if (this._isObject(source[key])) {
                    if (!target[key])
                        Object.assign(target, (_a = {}, _a[key] = {}, _a));
                    this._mergeDeep(target[key], source[key]);
                }
                else {
                    Object.assign(target, (_b = {}, _b[key] = source[key], _b));
                }
            }
        }
        return this._mergeDeep.apply(this, __spreadArrays([target], sources));
    };
    UI5BaseBuilder.prototype._enhanceWith = function (id, enhanceWith) {
        var oEnhanceBasis = {};
        if (typeof id === "string") {
            oEnhanceBasis.identifier.id = id;
        }
        else {
            oEnhanceBasis = this._mergeDeep(id, enhanceWith);
        }
        return oEnhanceBasis;
    };
    return UI5BaseBuilder;
}(UI5BaseBuilderIntf));
exports.UI5BaseBuilder = UI5BaseBuilder;
var UI5ChainSelection = /** @class */ (function (_super) {
    __extends(UI5ChainSelection, _super);
    function UI5ChainSelection() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    UI5ChainSelection.prototype.getThisPointer = function () {
        return this;
    };
    return UI5ChainSelection;
}(UI5BaseBuilder));
exports.UI5ChainSelection = UI5ChainSelection;
var UI5MessageToast = /** @class */ (function (_super) {
    __extends(UI5MessageToast, _super);
    function UI5MessageToast() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this._text = "";
        return _this;
    }
    UI5MessageToast.prototype.getThisPointer = function () {
        return this;
    };
    UI5MessageToast.prototype.format = function () {
        return "Message-Toast" + (this._text.length ? ("with text " + this._text) : "");
    };
    UI5MessageToast.prototype.withText = function (txt) {
        this._text = txt;
        return this;
    };
    UI5MessageToast.prototype.build = function () {
        if (this._text.length) {
            return testcafe_1.Selector(".sapMMessageToast").withText(this._text);
        }
        return testcafe_1.Selector(".sapMMessageToast");
    };
    return UI5MessageToast;
}(UI5BaseBuilder));
exports.UI5MessageToast = UI5MessageToast;
var UI5CoreItemSelection = /** @class */ (function (_super) {
    __extends(UI5CoreItemSelection, _super);
    function UI5CoreItemSelection() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    UI5CoreItemSelection.prototype.getThisPointer = function () {
        return this;
    };
    UI5CoreItemSelection.prototype.key = function (key) {
        return this.property("key", key);
    };
    return UI5CoreItemSelection;
}(UI5BaseBuilder));
exports.UI5CoreItemSelection = UI5CoreItemSelection;
var UI5ComboBoxChainSelection = /** @class */ (function (_super) {
    __extends(UI5ComboBoxChainSelection, _super);
    function UI5ComboBoxChainSelection() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    UI5ComboBoxChainSelection.prototype.getThisPointer = function () {
        return this;
    };
    UI5ComboBoxChainSelection.prototype.arrow = function () {
        return this.domChildWith("-arrow");
    };
    UI5ComboBoxChainSelection.prototype.item = function (key) {
        //merge own attributes and set as parent (to be improved)
        return ui5().parentId(this._id.identifier.id).coreItem().key(key);
    };
    return UI5ComboBoxChainSelection;
}(UI5BaseBuilder));
exports.UI5ComboBoxChainSelection = UI5ComboBoxChainSelection;
var UI5ListChainSelection = /** @class */ (function (_super) {
    __extends(UI5ListChainSelection, _super);
    function UI5ListChainSelection() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    UI5ListChainSelection.prototype.getThisPointer = function () {
        return this;
    };
    return UI5ListChainSelection;
}(UI5BaseBuilder));
exports.UI5ListChainSelection = UI5ListChainSelection;
var UI5TableRowChainSelection = /** @class */ (function (_super) {
    __extends(UI5TableRowChainSelection, _super);
    function UI5TableRowChainSelection() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    UI5TableRowChainSelection.prototype.getThisPointer = function () {
        return this;
    };
    UI5TableRowChainSelection.prototype.col0 = function () {
        return this.domChildWith("-col0");
    };
    return UI5TableRowChainSelection;
}(UI5BaseBuilder));
exports.UI5TableRowChainSelection = UI5TableRowChainSelection;
var UI5ObjectAttributeSelection = /** @class */ (function (_super) {
    __extends(UI5ObjectAttributeSelection, _super);
    function UI5ObjectAttributeSelection() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    UI5ObjectAttributeSelection.prototype.getThisPointer = function () {
        return this;
    };
    UI5ObjectAttributeSelection.prototype.inList = function (listId) {
        return this.parentId(listId);
    };
    UI5ObjectAttributeSelection.prototype.withText = function (text) {
        return this.property("text", text);
    };
    return UI5ObjectAttributeSelection;
}(UI5BaseBuilder));
exports.UI5ObjectAttributeSelection = UI5ObjectAttributeSelection;
function ui5(oChain) {
    return new UI5ChainSelection(oChain);
}
exports.ui5 = ui5;
