// ui5-testcafe-selector-utils/dist/ui5Config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ui5Config = void 0;
var ui5Password_1 = require("./ui5Password");
;
var ui5ConfigDef = /** @class */ (function () {
    function ui5ConfigDef() {
        try {
            this._config = require(process.cwd() + '\\.ui5-testcafe.json');
            if (typeof this._config.coverage === "undefined") {
                this._config.coverage = {};
            }
            if (typeof this._config.coverage.enabled === "undefined") {
                this._config.coverage.enabled = false;
            }
            if (typeof this._config.coverage.outDir === "undefined") {
                this._config.coverage.outDir = "/report/coverage";
            }
            if (typeof this._config.rest !== "undefined") {
                this._config.rest.password = ui5Password_1.ui5Password.decrypt(this._config.rest.password);
            }
            if (typeof this._config.launchpad === "undefined") {
                this._config.launchpad = {};
            }
            if (typeof this._config.launchpad.deactivateAnimation === "undefined") {
                this._config.launchpad.deactivateAnimation = false;
            }
        }
        catch (_a) {
            this._config = {
                coverage: {
                    outDir: "",
                    basePath: "",
                    type: "html",
                    enabled: false,
                    log: false,
                    debugComponents: [],
                    includePaths: [],
                    excludePaths: []
                },
                launchpad: {
                    deactivateAnimation: false
                },
                firstSelectorTimeout: 40000,
                traceSelectorOnFailure: false
            };
        }
        this._config.firstSelectorTimeout = this._config.firstSelectorTimeout ? this._config.firstSelectorTimeout : 40000;
        this._config.coverage.enabled = process.env.UI5_COVERAGE_ENABLED ? process.env.UI5_COVERAGE_ENABLED === "true" : this._config.coverage.enabled;
        this._config.traceSelectorOnFailure = typeof this._config.traceSelectorOnFailure === "undefined" ? false : this._config.traceSelectorOnFailure;
    }
    Object.defineProperty(ui5ConfigDef.prototype, "traceSelectorOnFailure", {
        get: function () {
            return this._config.traceSelectorOnFailure;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ui5ConfigDef.prototype, "firstSelectorTimeout", {
        get: function () {
            return this._config.firstSelectorTimeout;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ui5ConfigDef.prototype, "launchpad", {
        get: function () {
            return this._config.launchpad ? this._config.launchpad : { deactivateAnimation: false };
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ui5ConfigDef.prototype, "coverage", {
        get: function () {
            return this._config.coverage;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ui5ConfigDef.prototype, "rest", {
        get: function () {
            return this._config.rest;
        },
        enumerable: false,
        configurable: true
    });
    return ui5ConfigDef;
}());
exports.ui5Config = new ui5ConfigDef();
