// ui5-testcafe-selector-utils/dist/ui5Lumira.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ui5Lumira = void 0;
var testcafe_1 = require("testcafe");
var ui5Action_1 = require("./ui5Action");
var ui5Builder_1 = require("./ui5Builder");
var ui5Waiter_1 = require("./ui5Waiter");
;
var ui5LumiraDef = /** @class */ (function () {
    function ui5LumiraDef() {
    }
    ui5LumiraDef.prototype.startup = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var i;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.login(params.user, params.password)];
                    case 1:
                        _a.sent();
                        if (!params.afterLogin) return [3 /*break*/, 3];
                        return [4 /*yield*/, params.afterLogin()];
                    case 2:
                        _a.sent();
                        _a.label = 3;
                    case 3:
                        if (!params.parameter) return [3 /*break*/, 9];
                        i = 0;
                        _a.label = 4;
                    case 4:
                        if (!(i < params.parameter.length)) return [3 /*break*/, 7];
                        return [4 /*yield*/, this.enterParameter(params.parameter[i].number, params.parameter[i].value)];
                    case 5:
                        _a.sent();
                        _a.label = 6;
                    case 6:
                        i++;
                        return [3 /*break*/, 4];
                    case 7: return [4 /*yield*/, this.confirmParameter()];
                    case 8:
                        _a.sent();
                        return [3 /*break*/, 11];
                    case 9: return [4 /*yield*/, ui5Waiter_1.ui5Waiter.waitForBOToBeLoaded()];
                    case 10:
                        _a.sent();
                        _a.label = 11;
                    case 11: return [2 /*return*/];
                }
            });
        });
    };
    ui5LumiraDef.prototype.login = function (user, pw) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, ui5Action_1.ui5Action.typeText(ui5Builder_1.ui5().domQuery('#_id0\\3Alogon\\3AUSERNAME'), user)
                            .typeText(ui5Builder_1.ui5().domQuery('#_id0\\3Alogon\\3APASSWORD'), pw, { anonymize: true })
                            .click(ui5Builder_1.ui5().domQuery('#_id0\\3Alogon\\3AlogonButton'))];
                    case 1:
                        _a.sent();
                        //set focus on new iframe
                        return [4 /*yield*/, testcafe_1.t.switchToIframe("#openDocChildFrame")];
                    case 2:
                        //set focus on new iframe
                        _a.sent();
                        //wait 5 seconds for the i am loading from backend text to appear
                        return [4 /*yield*/, ui5Action_1.ui5Action.expect(ui5Builder_1.ui5().domQuery('#progressText')).exists().ok("Wait for Progress to be available", { timeout: 5000 })];
                    case 3:
                        //wait 5 seconds for the i am loading from backend text to appear
                        _a.sent();
                        return [4 /*yield*/, ui5Action_1.ui5Action.expect(ui5Builder_1.ui5().domQuery('#progressText')).exists().notOK("wait for Progress to disappear again", { timeout: 120000 })];
                    case 4:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    ui5LumiraDef.prototype.enterParameter = function (parameterNumber, parameterValue) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, ui5Action_1.ui5Action.typeText(ui5Builder_1.ui5().element(["sap.m.MultiInput", "sap.m.Input", "sap.m.DatePicker"])
                            .bindingContextPath(undefined, "/characteristics/" + parameterNumber.toString()), parameterValue, { replace: true }).
                            pressKey('enter')];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    ui5LumiraDef.prototype.confirmParameter = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, ui5Action_1.ui5Action.click(ui5Builder_1.ui5().id("OK"))];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, ui5Waiter_1.ui5Waiter.waitForBOToBeLoaded()];
                    case 2:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    return ui5LumiraDef;
}());
exports.ui5Lumira = new ui5LumiraDef();
